Add Queries object and consolidate queries there.

aynleslie opened the pull request:
#### What it do?

Much like we already have a `DocumentFilters` object to store our filters, I've added a `DocumentQueries` object to store our queries.  The queries where previously stored in 2 places:

- The DocumentClient
- The DocumentQueries inside the `Filters.scala` file

They are still in 2 places, with the following changes:

- The DocumentClient has only the ones required by the `BaseDocumentClient` and the now private `buildBaseRequest` method
- The DocumentQueries class has everything else

So:

- I moved a lot of queries
- I added the awful param `idsModRAStatuses`, a quadruple of Ints from the domainClient’s `calculateIdsAndModRAStatuses` method, to some of the query builders. this is an unfortunate consequence of me trying to break apart a huge PR and this piece is cleaner in the end, because that method will get moved.  It just doesn’t fit in at this phase (of breaking up the commit).   Please just tolerate this for a small while; I promise this is temporary.
- I did away with several of the wide sweeping `import com.socrata.cetera._` imports, and as such there are few small name changes.
- Two `scalastyle:ignore` commits were naturally done away with with these changes.   :)
- I added a bunch of new tests, so that all but 1 of the methods in the `Queries` file are tested. The one untested method is the `compositeFilteredQuery` and that is going to be refactored and simplified in the end, so it would be silly to write tests now. They'll be coming in a couple commits though. 

#### How was it tested?
sbt (all the usual things)
ran against prod, testing the /catalog endpoint with several param variations.

#### Warnings !!!

This is the stage-3 refactor PR. It is rebased on top of pull 171

jasonij commented on cetera-http/src/main/scala/com/socrata/cetera/search/DocumentClient.scala:
```diff
@@ -272,18 +85,12 @@ class DocumentClient(
+    val matchQuery = chooseMatchQuery(searchQuery, searchContext, fieldBoosts, datatypeBoosts,
+      minShouldMatch, slop, defaultTitleBoost, defaultMinShouldMatch)
 
+    val idsAndModRAStatuses = domainClient.calculateIdsAndModRAStatuses(domains)
```
I would linefeed after this.

jasonij replied:
And possibly add a comment about what's going on here. It's not clear exactly what this holds even though one can read what it says.

aynleslie replied:
as noted, this goes away in two more commits.  So really doesn't matter.

jasonij commented on cetera-http/src/main/scala/com/socrata/cetera/search/DocumentClient.scala:
```diff
@@ -6,6 +6,8 @@ import org.elasticsearch.search.aggregations.AggregationBuilders
 
 import com.socrata.cetera._
 import com.socrata.cetera.search.DocumentAggregations._
+import com.socrata.cetera.search.DocumentQueries.{compositeFilteredQuery, chooseMatchQuery}
```
Emergency! Emergency! Alphabetical order has been violated! Emergency! :dalek:

aynleslie replied:
shall fix.  Intellij's "optimize imports" is great.  Too bad it doesn't use the same logic when adding imports.  sigh.

jasonij commented on cetera-http/src/main/scala/com/socrata/cetera/search/DomainClient.scala:
```diff
@@ -7,7 +7,7 @@ import org.slf4j.LoggerFactory
 import com.socrata.cetera._
 import com.socrata.cetera.auth.CoreClient
-import com.socrata.cetera.search.DomainFilters.{domainIdsFilter, isCustomerDomainFilter}
+import com.socrata.cetera.search.DomainFilters.{idsFilter, isCustomerDomainFilter}
```
You're probably okay with just `customerDomainFilter`.

jasonij replied:
Oh never mind, that was already named that way.

jasonij commented on cetera-http/src/main/scala/com/socrata/cetera/search/Filters.scala:
```diff
@@ -216,10 +178,43 @@ object DocumentFilters {
+    val (domainIds, moderatedDomainIds, unmoderatedDomainIds, routingApprovalDisabledDomainIds) = idsModRAStatuses
+    val domainFilter = domainIdsFilter(domainIds)
+
+    val filter = FilterBuilders.boolFilter()
```
And then remove the FilterBuilders prefix here.

jasonij commented on cetera-http/src/main/scala/com/socrata/cetera/search/Filters.scala:
```diff
@@ -1,49 +1,11 @@
-import org.elasticsearch.index.query.MatchQueryBuilder.Type.PHRASE
-import org.elasticsearch.index.query.QueryBuilders._
-import org.elasticsearch.index.query.{FilterBuilder, QueryBuilder}
+import org.elasticsearch.index.query.{FilterBuilder, FilterBuilders}
```
Can clean this up:
```
import org.elasticsearch.index.query.FilterBuilder
import org.elasticsearch.index.query.FilterBuilders.{boolFilter, nestedFilter, notFilter, termFilter, termsFilter}
```

jasonij commented on cetera-http/src/main/scala/com/socrata/cetera/search/Filters.scala:
```diff
@@ -216,10 +178,43 @@ object DocumentFilters {
+
+  // TODO: (which is coming down the pike, soon I promise), remove need for that awful idsModRAStatuses param.
+  def compositeFilter(
+    domains: Set[Domain],
```
indent two more to the right

aynleslie replied:
aye aye.

jasonij commented on cetera-http/src/main/scala/com/socrata/cetera/search/Filters.scala:
```diff
@@ -216,10 +178,43 @@ object DocumentFilters {
   }
+
+  // TODO: (which is coming down the pike, soon I promise), remove need for that awful idsModRAStatuses param.
+  def compositeFilter(
```
Well, it *is* a composite filter. I think we had wanted, ultimately, to get the domain filtering factored out into its own function.

aynleslie replied:
yes, but this commit is only dealing with moving around Queries.

jasonij commented on cetera-http/src/main/scala/com/socrata/cetera/search/Filters.scala:
```diff
@@ -216,10 +178,43 @@ object DocumentFilters {
+    idsModRAStatuses: (Set[Int], Set[Int], Set[Int], Set[Int]))
+  : FilterBuilder = {
+
+    val contextModerated = searchContext.exists(_.moderationEnabled)
```
It took me a while to figure out what contextModerated meant. I might try another name like `isContextModerated` or `contextIsModerated` in a total reversal of my feelings about `is`. Or punt until later, just thinkin' cuz we're refactoring.

aynleslie replied:
I didn't rename this when I moved it.   We could, but in the interest of minimum delta I tried not mucking with the innards of the functions where I could avoid it.

jasonij replied:
I like the idea of minimum delta, at the same time we're so far beyond that. Generally, if I'm moving something anyway (so that the git history is lost) I might take the opportunity to clean it up.

Mostly I would say, since you have a chance to rename things while you're moving them, if a name strikes you as maybe warranting a change, see if you can come up with a better one.

jasonij commented on cetera-http/src/main/scala/com/socrata/cetera/search/Filters.scala:
```diff
@@ -216,10 +178,43 @@ object DocumentFilters {
+
+    val contextModerated = searchContext.exists(_.moderationEnabled)
+    val (domainIds, moderatedDomainIds, unmoderatedDomainIds, routingApprovalDisabledDomainIds) = idsModRAStatuses
+    val domainFilter = domainIdsFilter(domainIds)
```
It's weird to me that we grab the the 4 sets of domains, then define 1 filter from 1 of them up here, and create the others inside `List.concat ... foreach`. I would define vals for each of the domain-based filters that use these right here, then stick them into the list, so it's cleared how what they're being used for. Ideally there wouldn't be a lot of logic inside `List.concat`

aynleslie replied:
agree with you.  But again, I didn't write this, I just moved it around.  I'm happy to make this particular change in phase 6 though, since in that commit, I am changing the code quite a bit.

jasonij commented on cetera-http/src/main/scala/com/socrata/cetera/search/Filters.scala:
```diff
@@ -216,10 +178,43 @@ object DocumentFilters {
 
 object DomainFilters {
-  def domainIdsFilter(domainIds: Set[Int]): FilterBuilder = termsFilter("domain_id", domainIds.toSeq: _*)
+  def idsFilter(domainIds: Set[Int]): FilterBuilder = termsFilter("domain_id", domainIds.toSeq: _*)
```
You could probably call this `filterByIds` or something, which is consistent with Scala's use of filter, though honestly most people think filter means to take something away rather than select some things. So maybe `selectIds` or `selectById` or something, or just leave it, I just figure, well we have a chance to improve naming since we're moving everything around anyway.

aynleslie replied:
calling it with a verb-like name seems strange to me, since what we are creating is a noun, and in fact a Filter.  I only changed this because we do dumb things like `import com.socrata.cetera._` and documents also have a `domainIdsFilter` so things were colliding.  In order to be consistent with the way we've named filters, I called it `idsFilter` since it is already within the DomainFilters and the notion of Id is already scoped to Domains.

jasonij replied:
Okay sounds sensible, just thinking "out loud".

jasonij commented on cetera-http/src/main/scala/com/socrata/cetera/search/Queries.scala:
```diff
@@ -0,0 +1,219 @@
+
+object DocumentQueries {
+  private def applyDefaultTitleBoost(
+    defaultTitleBoost: Option[Float],
```
indent two more in here please (although I'm going to double check about the norm for handling private defs).

jasonij replied:
Yeah they still like two in.

aynleslie replied:
will do.

jasonij commented on cetera-http/src/main/scala/com/socrata/cetera/search/Queries.scala:
```diff
@@ -0,0 +1,219 @@
+    defaultTitleBoost: Option[Float],
+    defaultMinShouldMatch: Option[String],
+    fieldBoosts: Map[CeteraFieldType with Boostable, Float])
+  : Map[CeteraFieldType with Boostable, Float] = {
```
Incidentally, if you really don't like the look of:
```
def method(
    arg1: Type,
    arg2: Type)
  : Type = {
  code goes here
}
```

You can do
```
def method(
    arg1: Type,
    arg2: Type): Type = {
  code goes here
}
```

Just have to be consistent, but since you're moving so much, it's a chance to change it if you really prefer the latter.

aynleslie replied:
you are sweet to consider my distastes, but all of the rest of you like the former.  and I'll get used to it the more I see it.

jasonij commented on cetera-http/src/main/scala/com/socrata/cetera/search/Filters.scala:
```diff
@@ -216,10 +178,43 @@ object DocumentFilters {
+      Some(domainFilter), // TODO: remove me since I am the superset!
+      Some(publicFilter()),
+      Some(publishedFilter()),
+      Some(moderationStatusFilter(contextModerated, moderatedDomainIds, unmoderatedDomainIds)),
```
I'd really like to get these involved things "factored out" into vals with description names.
I'd also put the domain-related filters together in the list (without public and published in between them). I know it doesn't do anything but I think it increases readability to group things by similarity.

aynleslie replied:
again, these sorts of changes are coming in phase 6.   This change is primarily about moving methods into the appropriate classes.

jasonij commented on cetera-http/src/main/scala/com/socrata/cetera/search/Filters.scala:
```diff
@@ -216,10 +178,43 @@ object DocumentFilters {
+      Some(publishedFilter()),
+      Some(moderationStatusFilter(contextModerated, moderatedDomainIds, unmoderatedDomainIds)),
+      Some(routingApprovalFilter(searchContext, routingApprovalDisabledDomainIds)),
+      searchContext.flatMap(_ => domainMetadataFilter(domainMetadata)) // I make it hard to de-option
```
Ultimately I'd love to get rid of all the horrible optioning, but we can do that later I suppose.

aynleslie replied:
ya, and unfortunately, I have no phase coming for that.

jasonij commented on cetera-http/src/main/scala/com/socrata/cetera/search/Queries.scala:
```diff
@@ -0,0 +1,219 @@
+  }
+
+  def chooseMinShouldMatch(
+    minShouldMatch: Option[String],
```
two more indents

aynleslie replied:
yep yep

jasonij commented on cetera-http/src/main/scala/com/socrata/cetera/search/Queries.scala:
```diff
@@ -0,0 +1,219 @@
+  //
+  //   https://www.elastic.co/guide/en/elasticsearch/guide/current/proximity-relevance.html
+  def simpleQuery(
+    queryString: String,
```
two more space

aynleslie replied:
yep

jasonij commented on cetera-http/src/main/scala/com/socrata/cetera/search/Queries.scala:
```diff
@@ -0,0 +1,219 @@
+    slop: Option[Int],
+    defaultTitleBoost: Option[Float],
+    defaultMinShouldMatch: Option[String]
+  ): BaseQueryBuilder = {
```
align this, or put on the line above

jasonij replied:
well, ) belongs on line above, : is optional where it goes but if down here should be aligned.

jasonij commented on cetera-http/src/main/scala/com/socrata/cetera/search/Queries.scala:
```diff
@@ -0,0 +1,219 @@
+    searchQuery match {
+      case NoQuery => QueryBuilders.matchAllQuery
+      case AdvancedQuery(queryString) => advancedQuery(queryString, fieldBoosts)
+      case SimpleQuery(queryString) => simpleQuery(queryString,
```
Might add ( linefeed ... slop linefeed ) to desquish

jasonij commented on cetera-http/src/main/scala/com/socrata/cetera/search/Queries.scala:
```diff
@@ -0,0 +1,219 @@
+  : BaseQueryBuilder = {
+
+    // If there is no search context, use the ODN categories and tags
+    // otherwise use the custom domain categories and tags
```
I will always find this so confusing. We have got to go to being consistent with categories, tags, domain_categories, and domain_tags. There's no reason a public API should be unable to ask about things a site labelled 'Politics' which our classifier labelled 'Finance' or what have you.

Another time, but yeah, we gotta remember to do this.

jasonij commented on cetera-http/src/main/scala/com/socrata/cetera/search/Queries.scala:
```diff
@@ -0,0 +1,219 @@
+    // These constraints determine whether a document is considered part of the selection set, but
+    // they do not affect the relevance score of the document.
+    val compositeFilter = DocumentFilters.compositeFilter(
+      domains, datatypes, user, attribution, parentDatasetId, searchContext, domainMetadata, idsModRAStatuses)
```
Not greatly readable. I might try everything on its own line, or at least move the ) down.

aynleslie replied:
this goes down to two params in phase 6.  But sure can clean it up for now.

jasonij commented on cetera-http/src/main/scala/com/socrata/cetera/search/Queries.scala:
```diff
@@ -0,0 +1,219 @@
+    }
+  }
+
+  // scalastyle:ignore parameter.number
```
I think you need this on the same line, at least my scalastyle is objecting to the number of params.

jasonij commented on cetera-http/src/main/scala/com/socrata/cetera/search/Queries.scala:
```diff
@@ -0,0 +1,219 @@
+package com.socrata.cetera.search
+
+import org.elasticsearch.index.query.MatchQueryBuilder.Type._
+import org.elasticsearch.index.query.QueryBuilders._
```
Got to clean up these imports. Can't have QueryBuilders._ and QueryBuilder.

Either keep QueryBuilders._ and remove all the QueryBuilders. prefixes, or remove QueryBuilders._ and add the QueryBuilders prefix where necessary. I think it's okay to import wildcard and remove the QueryBuilders because the name collision issue seems not too great but either way.

aynleslie commented on cetera-http/src/main/scala/com/socrata/cetera/search/Queries.scala:
```diff
@@ -0,0 +1,219 @@
+      .should(QueryBuilders.hasParentQuery(esDomainType, domainQuery))
+  }
+
+  def chooseMatchQuery(
```
indents on all these.  I know.

jasonij commented on cetera-http/src/test/scala/com/socrata/cetera/search/DocumentClientSpec.scala:
```diff
@@ -794,7 +684,7 @@ class DocumentClientSpec extends WordSpec with ShouldMatchers with BeforeAndAfte
       }"""
 
-      val actual = documentClient.generateAdvancedQuery(
+      val actual = DocumentQueries.advancedQuery(
```
That's nice, so much more sense.

jasonij commented on cetera-http/src/test/scala/com/socrata/cetera/search/QueriesSpec.scala:
```diff
@@ -0,0 +1,332 @@
+    "return the expected query when nothing but a query is given" in {
+      val simpleQuery = DocumentQueries.simpleQuery(queryString, emptyFieldBoosts, emptyDatatypeBoosts, None, None)
+      val expected =
+        j"""{"bool" :
```
The indentation on the json interpolation is really messed up throughout this file. I think this should be fixed.

aynleslie replied:
please don't tell me you want it blown up to the complete "pretty" way.  I intentionally pulled things together to make it easier to read.  When everything is spread out across 30 lines,  I really can't take in the differences.  I tried to break it up into the meaningful chunks,  e.g I have one `bool` with  one `must` and one `should`.

jasonij replied:
Don't need to blow it to up pretty like, it's more readable a bit condensed, but we do want the parens aligned. It gets a little messier later on.

jasonij commented on cetera-http/src/test/scala/com/socrata/cetera/search/QueriesSpec.scala:
```diff
@@ -0,0 +1,332 @@
+
+  "DocumentQueries: compositeFilteredQuery" should {
+    "return the expected query if no categories are given" in {
+      // TODO:  I have a change forthcoming that simplifies this, so holding off on tests.
```
I would put a placeholder ignore so that we don't forget about this.

aynleslie replied:
sorry, don't follow.  isn't a TODO a reasonable way to remember to do something?

jasonij replied:
If you tell it to ignore a test, then you'll see "skipped!" or something when you run the test suite, so you don't forget to come back. TODOs just sit there quietly. Not super important but a useful practice I have found.

jasonij commented:
Lots of little things, most of which are discretionary. Imports should definitely be cleaned up (as this will also clean up the code base, as you'll see). The json interpolation indentation is quite messy and should be cleaned up.

Lots of indentation issues with fn signatures, I'm sure you can fix those quickly.

Variables names, well, worth having a thought about and deciding if you want to change them now or call them good. I do think in a couple places moving involved things out into their own vals would improve readability. Partly I see this as a chance, since we're moving the files around anyway, just to fix little things while we can and are looking at them. Up to you how much you want to do this.

jasonij commented on cetera-http/src/test/scala/com/socrata/cetera/search/QueriesSpec.scala:
```diff
@@ -0,0 +1,332 @@
+    "return the expected query when a query and field boosts are given" in {
+      val advancedQuery = DocumentQueries.advancedQuery(queryString, fieldBoosts)
+      val expected =
+        j"""{"bool" :{"should" :[
```
It's not the most important thing in the world, but you do want json blobs to be vaguely semanatic in their indentation. Here the root node is indented to the right of child nodes. Something like this:
```
      val expected = j"""
        {"bool" :{"should" :[
          {"query_string" :{"query" : "snuffy","fields" : [ "fts_analyzed", "fts_raw", "indexed_metadata.name^2.2"],"auto_generate_phrase_queries" : true}},
          {"has_parent" :{"query" :{"query_string" :{
            "query" : "snuffy",
            "fields" : ["fts_analyzed", "fts_raw", "domain_cname", "indexed_metadata.name^2.2"],
            "auto_generate_phrase_queries" : true}}, "parent_type" : "domain"}}
        ]}}
        """
```
(assuming that renders) is a big more readable to me because the indentation reflects the locations of the nodes.

jasonij replied:
Again not super crazy but just making sure a parent node is always to the left of a child node will keep it vaguely readable.

jasonij commented:
In general though, breaking things up into DocumentFilters and DocumentQueries seems a very sensible thing to do. We probably should have done that when we changed certain of our filters into queries. Also we might want to have put some comments in the code about why we did that because it's a bit subtle.

Oh well, I think it's very sensible and just needs some minor changes with optional other changes.

jasonij commented on cetera-http/src/test/scala/com/socrata/cetera/search/QueriesSpec.scala:
```diff
@@ -0,0 +1,332 @@
+            "query" : "snuffy",
+            "fields" : ["fts_analyzed", "fts_raw", "domain_cname"],
+            "auto_generate_phrase_queries" : true}}, "parent_type" : "domain"}
+          }]}}"""
```
Or for example here's a quick change that keeps the indentations semantic.
```
      val expected = j"""
        {"bool" :{"should" :[
          {"query_string" :{"query" : "snuffy","fields" : [ "fts_analyzed", "fts_raw" ],"auto_generate_phrase_queries" : true}},
          {"has_parent" :{"query" :{"query_string" :{
            "query" : "snuffy",
            "fields" : ["fts_analyzed", "fts_raw", "domain_cname"],
            "auto_generate_phrase_queries" : true}}, "parent_type" : "domain"}}
        ]}}"""
```
Just keeping parent to the left of child, and trying if I open { { { on the same line to close } } } on another same line.

jasonij commented on cetera-http/src/test/scala/com/socrata/cetera/search/QueriesSpec.scala:
```diff
@@ -0,0 +1,332 @@
+      val categoryQuery = DocumentQueries.categoriesQuery(Some(Set("Fun")))
+      val expected =
+        j"""{"nested": {"query": {"bool": {"should":
+          {"match": { "animl_annotations.categories.name": { "query": "Fun", "type": "phrase" }}},
```
Or indent the insides like four more to the right might be a quick fix.

aynleslie commented:
okay, so I've addressed everything save for the things that I have already changed on other branches. 
Can you take another look when you get a chance?

jasonij commented on cetera-http/src/main/scala/com/socrata/cetera/search/Filters.scala:
```diff
@@ -1,49 +1,11 @@
-import org.elasticsearch.index.query.QueryBuilders._
-import org.elasticsearch.index.query.{FilterBuilder, QueryBuilder}
+import org.elasticsearch.index.query.FilterBuilders.{boolFilter, nestedFilter, notFilter, termFilter, termsFilter}
+import org.elasticsearch.index.query.{FilterBuilder, FilterBuilders}
```
You don't need FilterBuilders, you've already imported all the filters from it you're going to need. Remove FilterBuilders and then remove it from the places where you're calling it explicitly because those methods are already there. Or, remove the top line and add `FilterBuilders.` where needed. I think it's better to have the filters you're using explicitly listed up top, fwiw. Then we know quickly and in one place all the filters we're actually using.

aynleslie replied:
oh, sorry, sorry.  Now I see what you are talking about.

jasonij commented on cetera-http/src/main/scala/com/socrata/cetera/search/Filters.scala:
```diff
@@ -216,10 +182,43 @@ object DocumentFilters {
+      Some(domainFilter), // TODO: remove me since I am the superset!
+      Some(publicFilter()),
+      Some(publishedFilter()),
+      Some(moderationStatusFilter(isContextModerated, moderatedDomainIds, unmoderatedDomainIds)),
```
We'll move these complicated constructs out into their own vals in a later PR?

aynleslie replied:
aye and I think you are really going to appreciate what is coming in phase 6.  stay tuned.

jasonij commented on cetera-http/src/main/scala/com/socrata/cetera/search/Queries.scala:
```diff
@@ -0,0 +1,226 @@
+    searchQuery match {
+      case NoQuery => QueryBuilders.matchAllQuery
+      case AdvancedQuery(queryString) => advancedQuery(queryString, fieldBoosts)
+      case SimpleQuery(queryString) => simpleQuery(queryString,
```
breathing room option here but would be nice (queryString next line, final paren new line)

jasonij commented:
Please clean up the imports in Filters.scala and then give it to the :sheep:

aynleslie commented:
sweet, sweet, sweet.  this is in.
